enhancements/test-platform/e2e-observer-pods.md

```diff
@@ -0,0 +1,115 @@
+
+## Summary
+
+e2e tests have multiple dimensions:
+ 1. which tests are running (parallel, serial, conformance, operator-specific, etc)
+ 2. which platform (gcp, aws, azure, etc)
+ 3. which configuration of that platform (proxy, fips, ovs, etc)
+ 4. which version is running (4.4, 4.5, 4.6, etc)
+ 5. which version change is running (single upgrade, double upgrade, minor upgrade, etc)
+Our individual jobs are an intersection of these dimensions and that intersectionality drove the development of steps
+to help handle the complexity.
+
+There is a kind of CI test observer that wants to run across all of these intersections by leveraging the uniformity of
+OCP across all the different dimensions above.
+It wants to do something like start an observer agent of some kind for every CI cluster ever created and provide o(100M)
+of data back to include in the CI job overall.
+Working with steps would require integrating in multiple dimensions and lead to gaps in coverage.
+
+## Motivation
+
+We need to run tools like
+ 1. e2e monitor - doesn't cover installation
+ 2. resourcewatcher - doesn't cover installation today if we wire it directly into tests.
+ 3. loki log collection (as created by group-b) - hasn't been able to work in upgrades at all.
+ 4. loki as provided by the logging team (future)
+To debug our existing install and upgrade failures.
+
+### Goals
+
+ 1. allow a cluster observing tool to run in the CI cluster (this avoid restarts during upgrades)
+ 2. allow a cluster observer to provide data (including junit) to be collected by the CI job
+ 3. collect data from cluster observer regardless of whether the job succeeded or failed
+ 4. allow multiple observers per CI cluster
+
+### Non-Goals
+
+ 1. allow a cluster observer to impact success or failure of a job
+ 2. allow a cluster observer to impact the running test.  This is an observer author failure.
+ 3. provide ANY dimension specific data. If an observer needs this, they need to integrate differently.
+ 4. allow multiple instances of a single cluster observer to run against one CI cluster
+
+## Proposal
+
+The overall goal: 
+ 1. have an e2e-observer process that is expected to be running before a kubeconfig exists
+ 2. the kubeconfig should be provided to the e2e-observer at the earliest possible time.  Even before it can be used.
+ 3. the e2e-observer process is expected to detect the presence of the kubeconfig itself
+ 4. the e2e-observer process will accept a signal indicating that teardown begins
+
+### Changes to CI
+This is a sketch of a possible path.
+ 1. Allow a non-dptp-developer to produces a manifest outside of any existing resource that can define
+    1. image
+    2. process
+    3. potentially a bash entrypoint
+    4. env vars
+    that mounts a `secret/<job>-kubeconfig` that will later contain a kubeconfig.
+    This happens to neatly align to a PodTemplate, but any file not tied to a particular CI dimension can work.
```

Comment:
2. Allow a developer to bind configured observer(s) to a job by one of the following mechanisms:
- attach the observer(s) to a step, so that any job which runs that step will run the observer
- attach the observer(s) to a workflow, so that any job which runs the workflow will run the observer
- attach the observer(s) to a literal test configuraiton, so that an observer could be added in a repo's test stanza
2a. Allow a developer to opt out of running named observer(s) by one of the following mechanisms:
- opt out of the observer(s) in a workflow, so that any job which runs the workflow will not run the observer
- opt out of the observer(s) in a literal test configuraiton, so that an observer could be excluded in a repo's test stanza

Reply:
That seems plausible.  I can't say that I'm very familiar with the mechanics available here.